pyfifotax/utils.py

```diff
@@ -0,0 +1,548 @@
+
+    return RawData(
+        df_rsu,
+        df_espp,
+        df_dividends,
+        df_buy_orders,
+        df_sell_orders,
+        df_currency_conversions,
+        df_stock_splits,
+    )
+
+
+def summarize_report(df_shares, df_forex, df_dividends, df_fees, df_taxes):
+    # this is simply the sum of all gains and losses
+    # for tax reasons, we usually also want the sum of gains
+    # and the sum of losses
+    share_gain_series = df_shares["Gain [EUR]"]
+    share_losses = sum_decimal(share_gain_series[share_gain_series < 0])
+    share_gains = sum_decimal(share_gain_series[share_gain_series > 0])
+
+    forex_gain_series = df_forex["Gain [EUR]"]
+
+    total_dividends = sum_decimal(df_dividends["Amount [EUR]"])
+    total_fees = sum_decimal(df_fees["Amount [EUR]"])
+    total_taxes = sum_decimal(df_taxes["Amount [EUR]"])
+
+    # unlike a previous version, we have to split things here
+    # losses from share can only be compared to gains from shares
+    # hence, there is a "total" gain/loss and then separate gains and
+    # losses from shares, e.g. one could have had a gain of 100
+    # and a loss of 150 over a year, i.e. a total loss of 50
+    total_foreign_gains = share_losses + share_gains + total_dividends
+    gains_from_shares = share_gains
+    losses_from_shares = -share_losses
+    total_gain_forex = sum_decimal(forex_gain_series)
+
+    anlagen = [
+        (
+            "Anlage KAP",
+            "Zeile 19: Ausländische Kapitalerträge (ohne Betrag lt. Zeile 47)",
+            round_decimal(total_foreign_gains),
+        ),
+        (
+            "Anlage KAP",
+            "Zeile 20: In den Zeilen 18 und 19 enthaltene Gewinne aus Aktienveräußerungen i. S. d. § 20 Abs. 2 Satz 1 Nr 1 EStG",
+            round_decimal(gains_from_shares),
+        ),
+        (
+            "Anlage KAP",
+            "Zeile 23: In den Zeilen 18 und 19 enthaltene Verluste aus der Veräuerung von Aktien i. S. d. § 20 Abs. 2 Satz 1 Nr. 1 EStG",
+            round_decimal(losses_from_shares),
+        ),
+        (
+            "Anlage KAP",
+            "Zeile 41: Anrechenbare noch nicht angerechnete ausländische Steuern",
+            round_decimal(total_taxes),
+        ),
+        (
+            "Anlage N",
+            "Zeile 48: (Werbungskosten Sonstiges): Überweisungsgebühren auf deutsches Konto für Gehaltsbestandteil RSU/ESPP",
+            round_decimal(total_fees),
+        ),
+        (
+            "Anlage SO",
+            "Zeilen 42 - 48: Gewinn / Verlust aus Verkauf von Fremdwährungen",
+            round_decimal(total_gain_forex),  # here: the sum should be fine
+        ),
+    ]
+    summary = {
+        "ELSTER - Anlage": [a[0] for a in anlagen],
+        "ELSTER - Zeile (Suggestion!)": [a[1] for a in anlagen],
+        "Value": [a[2] for a in anlagen],
+    }
+    df_summary = pd.DataFrame(summary)
+    return df_summary
+
+
+def create_report_sheet(name: str, df: pd.DataFrame, writer: pd.ExcelWriter):
+    if df.empty:
```

Comment:
This will never be empty because of the added commentaries. Need to find another way to skip the empty sheets.

Reply:
I am not sure if this actually was the case before as there might have been a row of empty cells anyways. Might have to check that.
On the other hand: technically it is actually preferable to always have the same sheets in the created reports, wouldn't you say?
I am tempted to remove this check completely.

Reply:
That's a good question. I myself liked it very much that I don't have to click through empty sheets just to see that it's, well, empty. For multiple years it's a real plus to see only those sheets which have actual content.

I also don't need to submit these empty PDF sheets to the tax office; and similarly WISO doesn't submit empty Anlage sheets either. So, I think the implicit meaning is: if something you don't declare that means that doesn't exist (e.g. renting out garage, selling house, buying a vacation house, etc.).

In case you'd like to add it back, I'd check whether it was empty before adding the extra summary lines.